src/gui/widgets/MicrotunerConfig.cpp

```diff
@@ -0,0 +1,648 @@
+			int num = 1, den = 1;
+			num = firstSection.toInt();
+			if (line.contains('/'))
+			{
+				den = line.split('/').at(1).section(QRegExp("\\s+"), 0, 0, QString::SectionSkipEmpty).toInt();
+			}
+			newIntervals.push_back(Interval(num, den));
+		}
+	}
+
+	Song *song = Engine::getSong();
+	if (song == NULL) {return false;}
+
+	auto newScale = std::make_shared<Scale>(m_scaleNameEdit->text(), newIntervals);
+	song->setScale(m_scaleComboModel.value(), newScale);
+
+	return true;
+}
+
+
+/**
+ * \brief Parse and apply the entered keymap definition
+ * \return true if input is valid, false if problems were detected
+ */
+bool MicrotunerConfig::applyKeymap()
+{
+	if (!validateKeymapForm()) {return false;}
+
+	std::vector<int> newMap;
+
+	QStringList input = m_keymapTextEdit->toPlainText().split('\n', QString::SkipEmptyParts);
+	for (auto &line: input)
+	{
+		if (line.size() == 0) {continue;}
+		if (line[0] == '!') {continue;}			// comment
+		QString firstSection = line.section(QRegExp("\\s+"), 0, 0, QString::SectionSkipEmpty);
+		if (firstSection == "x")
+		{
+			newMap.push_back(-1);				// not mapped
+			continue;
+		}
+		newMap.push_back(firstSection.toInt());
+	}
+
+	Song *song = Engine::getSong();
+	if (song == NULL) {return false;}
+
+	auto newKeymap = std::make_shared<Keymap>(m_keymapNameEdit->text(),
+											  newMap,
+											  m_firstKeyModel.value(),
+											  m_lastKeyModel.value(),
+											  m_middleKeyModel.value(),
+											  m_baseKeyModel.value(),
+											  m_baseFreqModel.value());
+	song->setKeymap(m_keymapComboModel.value(), newKeymap);
+
+	if (newKeymap->getDegree(newKeymap->getBaseKey()) == -1) {
+		QMessageBox::warning(this, tr("Invalid keymap"), tr("Base key is not mapped to any scale degree. No sound will be produced as there is no way to assign reference frequency to any note."));}
+
+	return true;
+}
+
+
+/**
+ * \brief Parse an .scl file and apply the loaded scale if it is valid
+ * \return true if input is valid, false if problems were detected
+ */
+bool MicrotunerConfig::loadScaleFromFile()
+{
+	QString fileName = FileDialog::getOpenFileName(this, tr("Open scale"), "", tr("Scala scale definition (*.scl)"));
+	if (fileName == "") {return false;}
+	QFile file(fileName);
+	if (!file.open(QIODevice::ReadOnly | QIODevice::Text))
+	{
+		QMessageBox::critical(this, tr("Scale load failure"), tr("Unable to open selected file."));
+		return false;
+	}
+	QTextStream stream(&file);
+	int i = -2, limit = 0;
+
+	m_scaleNameEdit->setText("");
+	m_scaleTextEdit->clear();
+	while (!stream.atEnd() && i < limit)
+	{
+		QString line = stream.readLine();
+		if (line != "" && line[0] == '!') {continue;}					// comment
+		switch(i) {
+			case -2:	m_scaleNameEdit->setText(line); break;			// first non-comment line = name or description
+			case -1:	limit = line.toInt(); break;					// second non-comment line = degree count
+			default:	m_scaleTextEdit->appendPlainText(line); break;	// all other lines = interval definitions
```

Comment:
```suggestion
		if (line != "" && line[0] == '!') {continue;}  // comment
		switch(i) {
			// first non-comment line = name or description
			case -2:	m_scaleNameEdit->setText(line); break;
			// second non-comment line = degree count
			case -1:	limit = line.toInt(); break;
			// all other lines = interval definitions
			default:	m_scaleTextEdit->appendPlainText(line); break;
```

Reply:
I don't know about that, I hate to waste so much vertical space (will be even worse in the keymap case..).